File: work-trial-starter/harness/trajectories/run_harness.py
#!/usr/bin/env python3
"""Correctness harness (Python). Green = every golden query implemented & correct.
    python3 harness/llm_spend/run_harness.py
Point at another adapter with ADAPTER=path/to/adapter.py (used by graders)."""
import json, os, sys, importlib.util

HERE = os.path.dirname(os.path.abspath(__file__))
golden = json.load(open(os.path.join(HERE, "golden_queries.json")))
adapter_path = os.path.abspath(os.environ.get("ADAPTER", os.path.join(HERE, "adapter.py")))
spec = importlib.util.spec_from_file_location("adapter", adapter_path)
mod = importlib.util.module_from_spec(spec)
spec.loader.exec_module(mod)


def grade(q, got):
    if got is None:
        return "PENDING"
    if q["result_type"] == "string[]":
        return "PASS" if list(got) == list(q["expected"]) else "FAIL"
    tol = q.get("tolerance_abs", 1e-6)
    return "PASS" if abs(float(got) - float(q["expected"])) <= tol + 1e-9 else "FAIL"


fail = pend = 0
for q in golden["queries"]:
    got, err = None, None
    try:
        got = mod.answer(q["id"])
    except Exception as e:  # noqa: BLE001
        err = str(e)
    state = grade(q, got)
    fail += state == "FAIL"
    pend += state == "PENDING"
    detail = "" if state == "PASS" else f"  got={('ERROR: ' + err) if err else got}  expected={q['expected']}"
    print(f"{state:<8} {q['id']}{detail}")

n = len(golden["queries"])
print(f"\n{n - fail - pend}/{n} passed · {fail} failed · {pend} pending")
sys.exit(1 if fail + pend > 0 else 0)

File: work-trial-starter/harness/llm_spend/run_harness.mjs
// Correctness harness (Node). Green = every golden query is implemented and correct.
//   node harness/llm_spend/run_harness.mjs
// Point it at a different adapter with ADAPTER=path/to/adapter.mjs (used by graders).
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const golden = JSON.parse(fs.readFileSync(path.join(HERE, "golden_queries.json"), "utf8"));
const adapterPath = process.env.ADAPTER
  ? path.resolve(process.cwd(), process.env.ADAPTER)
  : path.join(HERE, "adapter.mjs");
const { answer } = await import(pathToFileURL(adapterPath).href);

const fmt = (v) => (Array.isArray(v) ? `[${v.join(", ")}]` : String(v));
function grade(q, got) {
  if (got === null || got === undefined) return "PENDING";
  if (q.result_type === "string[]") return JSON.stringify(got) === JSON.stringify(q.expected) ? "PASS" : "FAIL";
  const tolAbs = q.tolerance_abs ?? 1e-6;
  return Math.abs(Number(got) - Number(q.expected)) <= tolAbs + 1e-9 ? "PASS" : "FAIL";
}

let fail = 0, pend = 0;
for (const q of golden.queries) {
  let got = null, err = null;
  try { got = await answer(q.id); } catch (e) { err = e.message; }
  const state = grade(q, got);
  if (state === "FAIL") fail++;
  if (state === "PENDING") pend++;
  const detail = state === "PASS" ? "" : `  got=${err ? "ERROR: " + err : fmt(got)}  expected=${fmt(q.expected)}`;
  console.log(`${state.padEnd(8)} ${q.id}${detail}`);
}
const passed = golden.queries.length - fail - pend;
console.log(`\n${passed}/${golden.queries.length} passed · ${fail} failed · ${pend} pending`);
process.exit(fail + pend > 0 ? 1 : 0);
